uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php

```diff
@@ -910,6 +910,219 @@ public function testManagerCanCancelOrRejectLeaveRequestEvenIfBalanceIsGreaterTh
     }
   }
 
+  public function testManagerCanNotApproveLeaveRequestIfBalanceChangeIsMoreThanEntitlementBalanceWhenAllowOveruseFalse() {
```

Comment:
The tests you added all mention a manager or a staff, but I don't think this is really necessary. The validation you're trying to test here doesn't care about the user role. At this point, we should only worry about the balances. 

What we need to test here is that an open leave request cannot be updated if its balance change will make the balance negative). In this case an open leave request can be either:
- An Approved Leave Request that was updated to either Awaiting approval or More Information Required
- An More Information Required that was updated to Awaiting Approval or vice versa

Reply:
@davialexandre so you mean it should be possible (for example) transit between More Information Required and Awaiting Approval even if the closing balance is negative?

Reply:
Sure. Just wondering, is such an approach allowed? It allows to test multiple transitions in one test:
```js
$statusTransitions = [
  ['approved', 'more_information_required'],
  ['awaiting_approval', 'approved']
];

foreach ($statusTransitions as $statusTransition) {
  $statusFrom = $statusTransition[0];
  $statusTo = $statusTransition[1];
  $exceptionMessage = '';

  // Balance change is 1 since this is a single day leave request
  $leaveRequestData = [
    'type_id' => $absenceType->id,
    'contact_id' => $periodEntitlement->contact_id,
    'status_id' => $statusFrom,
    'from_date' => $requestDate,
    'from_date_type' => $requestDateType,
    'to_date' => $requestDate,
    'to_date_type' => $requestDateType,
    'request_type' => LeaveRequest::REQUEST_TYPE_LEAVE
  ];

  // Create a new leave request
  $leaveRequest = LeaveRequestFabricator::fabricateWithoutValidation($leaveRequestData, true);

  // Attempt to change leave request status
  $updatedLeaveRequestData = $leaveRequestData;
  $updatedLeaveRequestData['id'] = $leaveRequest->id;
  $updatedLeaveRequestData['status_id'] = $statusTo;

  // Attempt to transit the leave request status
  try {
    $updatedLeaveRequest = LeaveRequest::create($updatedLeaveRequestData);
  } catch (Exception $e) {
    $exceptionMessage = $e->getMessage();
  }

  $this->assertEquals($exceptionMessage, 'There 2 are only '. $entitlementBalance .' days leave available. This request cannot be made or approved');
}
```